### PR TITLE
Hd/feature extraction

### PR DESCRIPTION
**Pull request by kennymckormick**
Yield TSN Feature for BSN / BMN

### CODE REVIEW

**Comment by dreamerlin** [comment]
In this repo, we use `activitynet` instead of `anet`.

File: configs/recognition/tsn/activitynet/tsn_r50_320p_1x1x8_150e_anet_clip_flow.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+# model training and testing settings
+train_cfg = None
+test_cfg = dict(average_clips=None)
+# dataset settings
+dataset_type = 'RawframeDataset'
+data_root = 'data/anet/flow'
+data_root_val = 'data/anet/flow'
+ann_file_train = 'data/anet/anet_train_clip.txt'
+ann_file_val = 'data/anet/anet_val_clip.txt'
+ann_file_test = 'data/anet/anet_val_clip.txt'
```
Comment (dreamerlin) [inline configs/recognition/tsn/activitynet/tsn_r50_320p_1x1x8_150e_anet_clip_flow.py:27]:
```
data_root = 'data/ActivityNet/flow'
data_root_val = 'data/ActivityNet/flow'
ann_file_train = 'data/ActivityNet/anet_train_clip.txt'
ann_file_val = 'data/ActivityNet/anet_val_clip.txt'
ann_file_test = 'data/ActivityNet/anet_val_clip.txt'
```
To keep consist with [this](https://github.com/open-mmlab/mmaction2/blob/master/configs/localization/bmn/bmn_400x100_2x8_9e_activitynet_feature.py#L17-L22)

File: configs/recognition/tsn/activitynet/tsn_r50_320p_1x1x8_150e_anet_clip_flow.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
+# model settings
```
Comment (dreamerlin) [inline configs/recognition/tsn/activitynet/tsn_r50_320p_1x1x8_150e_anet_clip_flow.py:1]: `tsn_r50_320p_1x1x8_150e_activitynet_clip_flow.py`

File: tools/anet_feature_prepare.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+    return args
+
+
+def pool_feature(data, num_prop=100, num_sample_bin=3, pool_type='mean'):
+    """Pool features with arbitrary temporal length.
+
+    Args:
+        data (list[np.ndarray] or np.ndarray): features of an untrimmed video,
+            with arbitrary temporal length.
+        num_prop (int): The temporal dim of pooled feature. Default: 100.
```
Comment (dreamerlin) [inline tools/anet_feature_prepare.py:29]: `num_prop` -> `num_proposals`

File: tools/anet_feature_prepare.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+        num_sample_bin (int): How many points to sample to get the feature
```
Comment (dreamerlin) [inline tools/anet_feature_prepare.py:30]: `num_sample_bin` -> `num_sample_bins`

File: tools/anet_feature_prepare.py [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+    feat = np.concatenate([rgb_feat, flow_feat], axis=-1)
+    if args.output_format == 'pkl':
+        dump(feat, osp.join(args.dest, name))
+    elif args.output_format == 'csv':
+        feat = feat.tolist()
+        lines = []
+        line0 = ','.join(['f{}'.format(i) for i in range(400)])
+        lines.append(line0)
+        for line in feat:
+            lines.append(','.join(['{:.4f}'.format(x) for x in line]))
```
Comment (dreamerlin) [inline tools/anet_feature_prepare.py:81]:
```
lines.append(','.join(['f{x:.4f}' for x in line]))
```

File: tools/anet_feature_prepare.py [hunk 6]
```diff
@@ -0,0 +1,97 @@
...
+    flow_feat = load(osp.join(args.flow, name))
+    rgb_feat = pool_feature(rgb_feat)
+    flow_feat = pool_feature(flow_feat)
+    feat = np.concatenate([rgb_feat, flow_feat], axis=-1)
+    if args.output_format == 'pkl':
+        dump(feat, osp.join(args.dest, name))
+    elif args.output_format == 'csv':
+        feat = feat.tolist()
+        lines = []
+        line0 = ','.join(['f{}'.format(i) for i in range(400)])
```
Comment (dreamerlin) [inline tools/anet_feature_prepare.py:78]:
```
line0 = ','.join(['f{i}' for i in range(400)])
```
and is `f` a typo here?

File: tools/tsn_feature_extract.py [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+from mmaction.models import build_model
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Extract TSN Feature')
+    parser.add_argument('--data-prefix', default='', help='dataset prefix')
+    parser.add_argument('--output-prefix', default='', help='output prefix')
+    parser.add_argument(
+        '--data-list',
+        help='video list of the \
```
Comment (dreamerlin) [inline tools/tsn_feature_extract.py:19]: Avoid to use `\`, just renew a line with `''`
```
help='video list of the dataset, the format should be '
     '`frame_dir num_frames output_file`'
```

File: tools/tsn_feature_extract.py [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+        sample = data_pipeline(tmpl)
+        imgs = sample['imgs']
+        shape = imgs.shape
+        # the original shape should be N_seg * C * H * W, resize it to N_seg *
+        # 1 * C * H * W so that the network return feature of each frame (No
+        # score average among segments)
+        imgs = imgs.reshape((
+            shape[0],
+            1,
+        ) + shape[1:])
```
Comment (dreamerlin) [inline tools/tsn_feature_extract.py:120]:
```
imgs = imgs.reshape((shape[0], 1) + shape[1:])
```

File: tools/tsn_feature_extract.py [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+    model.eval()
+
+    data = open(args.data_list).readlines()
+    data = [x.strip() for x in data]
+    data = data[args.part::args.tot]
+
+    # enumerate Untrimmed videos, extract feature from each of them
+    for item in tqdm(data):
+        item = item.split()
+        frame_dir, length, output_file = item
```
Comment (dreamerlin) [inline tools/tsn_feature_extract.py:99]: frame_dir, length, output_file = item.split()

File: tools/anet_feature_prepare.py [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+    parser.add_argument('--output-format', default='pkl', help='clip length')
+    args = parser.parse_args()
+    return args
+
+
+def pool_feature(data, num_prop=100, num_sample_bin=3, pool_type='mean'):
+    """Pool features with arbitrary temporal length.
+
+    Args:
+        data (list[np.ndarray] or np.ndarray): features of an untrimmed video,
```
Comment (dreamerlin) [inline tools/anet_feature_prepare.py:27]: `data (list[np.ndarray] | np.ndarray)`

Comment (dreamerlin) [reply tools/anet_feature_prepare.py:27]: feature -> Feature

File: tools/tsn_feature_extract.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    model.load_state_dict(state_dict)
+    model = model.cuda()
+    model.eval()
+
+    data = open(args.data_list).readlines()
+    data = [x.strip() for x in data]
+    data = data[args.part::args.tot]
+
+    # enumerate Untrimmed videos, extract feature from each of them
+    for item in tqdm(data):
```
Comment (dreamerlin) [inline tools/tsn_feature_extract.py:97]:
```
prog_bar = mmcv.ProgressBar(len(data))
for item in data:
    ...
    ...
    prog_bar.update()
```

File: tools/anet_feature_prepare.py (hunk 6 above)
Comment (kennymckormick) [reply tools/anet_feature_prepare.py:78]: no, just follow the format of existing features

**Pull request closed without merging**